### ISSUE TITLE
Pointer debug framework in pgexporter

### ISSUE THREAD

**Issue by resyfer** [pgexporter/pgexporter#107]
Since most of the bugs are memory-related bugs, and which involve invalid or NULL pointers, I feel it would be very important, given the current complexity of the code, to have a pointer debug "framework" inside pgexporter.

I was thinking along the lines of a function like:
```c
#ifdef DEBUG
void debug_ptr(void *ptr, size_t bytes, char flags, char *ptr_name);
#endif
```

That would check for permission according to flags provided (`PTR_READ` for read check, `PTR_WRITE` for write check and `PTR_READ | PTR_WRITE` for read/write check, and so on for free, unmap, etc.).

Though a proper framework would be a good choice, my personal suggestion would be to have a simple sandbox using fork, perform a small memory operation (read operation if read, and so on) on the child at that pointer, and print the required error message if the child crashes due to the operation, and also prints some information using the `ptr_name` including the range of addresses (can catch NULL) as well as which of the flags provided are allowed (for cases like mmap for private or read-only memory).

```
Name      Range            Size Flags Allowed            Flags Not Allowed

testptr   NULL             1    PTR_READ, PTR_WRITE      -
testptr2  0x7fff5ed98c4c   4    PTR_READ                 PTR_WRITE
```
something along those lines.

Any write operation(s) in child will cause [Copy On Write](https://en.wikipedia.org/wiki/Copy-on-write), so not an issue of changing original pointer(s).

**Comment by jesperpedersen** [comment]
`pgexporter_log_mem(void* data, size_t size);` ?

**Comment by resyfer** [comment]
This _can_ work as an improvement to the logging function, but I intended this to be more focused on the info about the read/write permissions on the pointer (includes use-after-free). The output of range was included primarily to provide information about NULL pointers.

I also intended this (both feature and the info it gives) to be strictly available during development (ie. for developers), as it is a very slow (though effective) method of testing pointers, but I don't think it would be a problem if it's publicly available.

While improving the log feature would work without issues, these are some points that make me lean slightly towards making another function.

**Comment by jesperpedersen** [comment]
Well, start with a `#if DEBUG` function, and we can take it from there

**Comment by sunilkumar2170** [comment]
Hi @resyfer 👋  

I’d like to work on this issue.

I’m planning to start with a minimal `#if DEBUG` implementation of a
`pgexporter_debug_ptr()` helper that:
- detects NULL pointers
- checks read/write access using a fork-based sandbox
- logs clear diagnostic messages using existing pgexporter logging

I’ll keep the initial version simple and developer-only, and we can iterate
based on feedback.

If this approach sounds good, I’ll open a draft PR soon. Thanks!

**Comment by jesperpedersen** [comment]
@sunilkumar2170 You should start by investigating if clang/ASAN when using 21+ is enough before starting on an implementation

**Comment by sunilkumar2170** [comment]
Thanks for the guidance @jesperpedersen 🙏  

Understood. I won’t start implementing anything yet. I’ll first investigate
whether clang + AddressSanitizer (ASAN), especially with clang 21+, already
covers the pointer-related issues discussed here (NULL dereference,
use-after-free, invalid read/write).

I’ll build pgexporter with ASAN enabled, test relevant cases, and share my
findings here. Based on that, we can decide if a custom DEBUG-only pointer
debug helper is still necessary.

Thanks!

**Comment by Abdelrhmansersawy** [comment]
@jesperpedersen 
I believe this feature is very important for pgexporter, and I’m interested in working on it if no one else is currently assigned. Could you please assign it to me?

I’ll start with thorough research and design work alongside the issue I’m currently working on. Once that’s finished, I’ll open a draft PR presenting the proposed design for the debug framework.

**Comment by resyfer** [comment]
@Abdelrhmansersawy this is a feature already assigned to @sunilkumar2170. @sunilkumar2170 do you have any updates on this? Or do you think a reassignment of the issue is alright?

**Comment by Abdelrhmansersawy** [comment]
> [@Abdelrhmansersawy](https://github.com/Abdelrhmansersawy) this is a feature already assigned to [@sunilkumar2170](https://github.com/sunilkumar2170). [@sunilkumar2170](https://github.com/sunilkumar2170) do you have any updates on this? Or do you think a reassignment of the issue is alright?

Yes, there haven’t been any updates from him for the past two weeks, so I wanted to check whether he’s still working on it. If not I can start working on it.

**Comment by sunilkumar2170** [comment]
Hi @resyfer, thanks for checking.

Sorry for the delay. I am still working on this issue.

I am currently investigating clang + ASAN (especially clang 21+) to see what classes of pointer-related bugs are already detected. I am setting up builds and test cases and will share my findings soon.

I would like to continue owning this issue.

Thanks!

**Comment by jesperpedersen** [comment]
@sunilkumar2170 Keep at it - clang 22 will be out Feb 24 - save some time by trying the -rc releases

**Comment by sunilkumar2170** [comment]
Thanks! I’ll try the clang 22 RC builds as well and see how ASAN behaves there. Will share updates soon.
